allennlp/semparse/domain_languages/domain_language.py

```diff
@@ -0,0 +1,474 @@
+        return_nltk_type = BasicType(return_type)
+        function_nltk_type = PredicateType.get_function_type(argument_nltk_types, return_nltk_type)
+        self._functions[name] = function
+        self._function_types[name] = function_nltk_type
+
+    def add_constant(self, name: str, value: Any):
+        """
+        Adds a constant to this domain language.  You would typically just pass in a list of
+        constants to the ``super().__init__()`` call in your constructor, but you can also call
+        this method to add constants if it is more convenient.
+
+        Because we construct a grammar over this language for you, in order for the grammar to be
+        finite we cannot allow arbitrary constants.  Having a finite grammar is important when
+        you're doing semantic parsing - we need to be able to search over this space, and compute
+        normalized probability distributions.
+        """
+        constant_type = BasicType(type(value))
+        self._functions[name] = lambda: value
+        self._function_types[name] = constant_type
+
+    def _execute_expression(self, expression: Any):
+        """
+        This does the bulk of the work of executing a logical form, recursively executing a single
+        expression.  Basically, if the expression is a function we know about, we evaluate its
+        arguments then call the function.  If it's a list, we evaluate all elements of the list.
+        If it's a constant (or a zero-argument function), we evaluate the constant.
+        """
+        # pylint: disable=too-many-return-statements
+        if isinstance(expression, list):
+            if isinstance(expression[0], list):
+                function = self._execute_expression(expression[0])
+            elif expression[0] in self._functions:
+                function = self._functions[expression[0]]
+            else:
+                if isinstance(expression[0], str):
+                    raise ExecutionError(f"Unrecognized function: {expression[0]}")
+                else:
+                    raise ExecutionError(f"Unsupported expression type: {expression}")
+            arguments = [self._execute_expression(arg) for arg in expression[1:]]
+            try:
+                return function(*arguments)
+            except (TypeError, ValueError) as error:
+                raise ExecutionError(f"Error executing expression {expression}: {error}")
+        elif isinstance(expression, str):
+            if expression not in self._functions:
+                raise ExecutionError(f"Unrecognized constant: {expression}")
+            try:
+                return self._functions[expression]()
+            except (TypeError, ValueError) as error:
+                raise ExecutionError(f"Error executing expression {expression}: {error}")
+        else:
+            raise ExecutionError("Not sure how you got here. Please open a github issue with details.")
```

Comment:
I feel like there's a cleaner way (e.g. fewer cases) to write this function, but I've been staring at it for 5 minutes and can't figure out what it is.

Reply:
I think if we just assumed good input, or let the code crash on bad input without switching to an `ExecutionError`, we could make it a lot simpler.  Trying to be nice about handling bad input is what makes it so complicated.